// Copyright Â© 2017 Canonical Ltd.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License version 3 as
// published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// Authored by: Alberto Aguirre <alberto.aguirre@canonical.com>

syntax = "proto3";
package multipass;

service Rpc {
    rpc version (VersionRequest) returns (VersionReply);
    rpc launch (LaunchRequest) returns (LaunchReply);
    rpc ssh (SSHRequest) returns (SSHReply);
    rpc start (StartRequest) returns (StartReply);
    rpc stop (StopRequest) returns (StopReply);
    rpc destroy (DestroyRequest) returns (DestroyReply);
    rpc list (ListRequest) returns (ListReply);
}

message VersionRequest {
}

message VersionReply {
    string version = 1;
}

message LaunchRequest {
    string vm_name = 1;
    string release = 2;
    string kernel_name = 3;
    int32 num_cores = 4;
    int32 mem_size = 5; // units?
    int32 disk_space = 6; // units?
    //string cloud_init_config_path = 7;
}

message LaunchReply {
    string vm_instance_name = 1;
}

message SSHRequest {
    string vm_name = 1;
}

message SSHReply {
}

message StartRequest {
    string vm_name = 1;
}

message StartReply {
}

message StopRequest {
    string vm_name = 1;
}

message StopReply {
}

message DestroyRequest {
    string vm_name = 1;
    bool force = 2;
}

message DestroyReply {
}

message ListRequest {
}

message ListReply {
    repeated string vm_instance_names = 1;
}
